### PR TITLE
[FEAT][ROCm] Upgrade AITER MLA v1 backend

### PR DESCRIPTION
**Pull request by vllmellm**
This PR upgrades the AITER MLA attention backend on the v1 engine, as this backend was previously upgraded on the v0 engine in [this PR](https://github.com/vllm-project/vllm/pull/17912). The upgrade uses AITER commit [c1debd8](https://github.com/ROCm/aiter/commit/c1debd87ce0391aa27438d9e07e76e4fea7c4b70), which introduces the relevant API changes.

It's important to note that the `mla_decode_fwd` kernel in the AITER package imposes new constraints on supported attention heads, requiring either 128 or 16 heads. To load DeepSeek models correctly, you must configure the `tensor_parallel_size` parameter so that each rank processes 16 heads. 

For example, on MI300X AMD devices, the following configurations are required:

- DeepSeek-v3/DeepSeek-v2 (128 total heads): Use tensor_parallel_size=8 (128 ÷ 16 = 8 ranks)

 - DeepSeek-v2-Lite (16 total heads): Use tensor_parallel_size=1 (16 ÷ 16 = 1 rank)

### deepseek-ai/DeepSeek-V3 on v1 engine
`VLLM_USE_V1=1
VLLM_ATTENTION_BACKEND="ROCM_AITER_MLA"
SAFETENSORS_FAST_GPU=1
lm_eval --model vllm --model_args pretrained=deepseek-ai/DeepSeek-V3,tensor_parallel_size=8,max_model_len=32768,block_size=1 --trust_remote_code --tasks gsm8k --num_fewshot 5 --batch_size auto`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9439|±  |0.0063|
|     |       |strict-match    |     5|exact_match|↑  |0.9439|±  |0.0063|

### CODE REVIEW

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 1]
```diff
@@ -189,7 +202,18 @@ def _forward_decode(
...
         kv_buffer = kv_c_and_k_pe_cache.unsqueeze(2)
 
+        if self.num_heads == 16:
+            # AITER MLA decode kernel only supports
+            # max_seqlen_q=1 when using 16 heads.
+            max_seqlen_qo = 1
+        else:
+            # AITER MLA decode Kernel handles arbitrary
+            # max_seqlen_q values when using 128 heads.
+            assert attn_metadata.prefill is not None
```
Comment (ProExpertProg) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py:212]: Is this always set?

**Review by ProExpertProg — COMMENTED** [review]
A few questions

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 2]
```diff
@@ -189,7 +202,18 @@ def _forward_decode(
 
         kv_buffer = kv_c_and_k_pe_cache.unsqueeze(2)
 
+        if self.num_heads == 16:
+            # AITER MLA decode kernel only supports
+            # max_seqlen_q=1 when using 16 heads.
+            max_seqlen_qo = 1
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/mla/rocm_aiter_mla.py:208]: should we assert here? seems dangerous to just change this since it may cause a mismatch with `qo_indptr`?

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py [hunk 3]
```diff
@@ -75,27 +77,33 @@ def _get_paged_kv_tensors(
...
-                        dtype=block_table_bounds.dtype,
-                        device=block_table_bounds.device),
+            torch.zeros(1, dtype=block_table_bounds.dtype, device=device),
             block_table_bounds.cumsum(dim=0, dtype=torch.int32)
         ])
 
         paged_kv_last_page_len = seq_lens % page_size
         paged_kv_last_page_len = torch.where(paged_kv_last_page_len == 0,
                                              page_size, paged_kv_last_page_len)
+        qo_indptr = torch.arange(0,
```
[the remark this answers is not in the document]
Comment (vllmellm) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:96]: @ProExpertProg it's a pointer array used to manage query sequences in AITER `mla_deocde_fwd ` kernel. The variable name is used here is same as it is name to it in AITER kernel ([see more](https://github.com/ROCm/aiter/blob/adfca087942fca2c4afb8b8e6d65063b02a4166f/aiter/mla.py#L98)).

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 2 above)
Comment (vllmellm) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:208]: @LucasWilkinson in decode forward ideally the kernel processes one token at a time per sequence and is not constraint to length of `qo_indptr` which is a pointer array used to manage decode query sequences, this allows identifying the start and end of decode query sequence ([see more](https://github.com/ROCm/aiter/blob/adfca087942fca2c4afb8b8e6d65063b02a4166f/aiter/mla.py#L36-L38)). While `max_seqlen_q` determines the grid dimension for parallel processing: grid = (bs, nhead, max_seqlen_q=1) which invokes the kernel using this grid size ([see more](https://github.com/ROCm/aiter/blob/adfca087942fca2c4afb8b8e6d65063b02a4166f/aiter/mla.py#L161)) and it is constraint to the number of heads to be 16 only which means processes one token at a time during decode. this is a hard constraint set in the kernel ([see more](https://github.com/ROCm/aiter/blob/adfca087942fca2c4afb8b8e6d65063b02a4166f/aiter/mla.py#L124)).

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 1 above)
Comment (vllmellm) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:212]: @ProExpertProg as the metadata required here is `max_seqlen_q` which is stored part of prefill metadata I think it would be better if we could modify the common module and store this in common metadata so we wouldn't need to access this through prefill. Generally speaking `max_query_len` is passed `build` function as below in common module:

https://github.com/vllm-project/vllm/blob/0c15c2e4868173642cec766c9819a210aef5e518/vllm/v1/attention/backends/mla/common.py#L452-L454

https://github.com/vllm-project/vllm/blob/0c15c2e4868173642cec766c9819a210aef5e518/vllm/v1/attention/backends/mla/common.py#L539-L544

or we could simply pass this argument to `build_decode` function such that the function would determine its own `max_decode_query_len` metadata.

https://github.com/vllm-project/vllm/blob/0c15c2e4868173642cec766c9819a210aef5e518/vllm/v1/attention/backends/mla/common.py#L445-L450

perhaps @LucasWilkinson have better ideas to suggest here as well. would appreciate your suggestion.

File: vllm/v1/attention/backends/mla/rocm_aiter_mla.py (hunk 2 above)
Comment (LucasWilkinson) [reply vllm/v1/attention/backends/mla/rocm_aiter_mla.py:208]: my preference would be to store a separate `max_seqlen_qo` for decode (we do this for most attention backends that split prefill decode) but thats just personal opinion

**Review by LucasWilkinson — APPROVED** [review]
LGTM once green (we may need a force merge but lets wait for all the tests to finish)

**Comment by DarkLight1337** [comment]
Merging despite the PR freeze, since this only changes ROCm code path
